kernel/src/main/scala/cats/kernel/instances/FunctionInstances.scala

```diff
@@ -106,11 +108,26 @@ private[instances] trait FunctionInstances4 {
     new Function1Semigroup[A, B] { def B: Semigroup[B] = S }
 }
 
+final private[instances] case class CombineFunction1[A, B](left: A => B, right: A => B, semiB: Semigroup[B])
+    extends (A => B) {
+  private def call(fn: A => B, a: A): TailRec[B] =
+    fn match {
+      case CombineFunction1(l, r, sg) =>
+        for {
+          lb <- tailcall(call(l, a))
+          rb <- tailcall(call(r, a))
+        } yield sg.combine(lb, rb)
```

Comment:
looks like scala 3 doesn't like this line. I think we could possibly fix it by doing:
```scala
private def call[C](fn: A => C, a: A): TailRec[C] =
```
and not tie the result type to `B`. Since `Semigroup` is invariant in `B` but `A => B` is covariant in B, actually the pattern match could give a `B1 <: B`. I think `TailRec` is covariant, so this should be fine... maybe scala 3 has a bug here, or I'm missing something...

Reply:
That fixes it in scala 3 but breaks it in scala 2 😢

```
[error] FunctionInstances.scala:119:28: type mismatch;
[error]  found   : lc.type (with underlying type C)
[error]  required: ?C1 where type ?C1 <: C (this is a GADT skolem)
[error]         } yield sg.combine(lc, rc)
```

What I found does work is using a type ascription in the match along with an `@unchecked` annotation, similar to [what `AndThen` does here](https://github.com/typelevel/cats/blob/main/core/src/main/scala/cats/data/AndThen.scala#L143).

I went ahead a committed that but would be happy to revert it if there's a better solution.

Reply:
This seems to work for me:
```scala
final case class CombineFunction1[-A, B](left: A => B, right: A => B, semiB: Semigroup[B]) extends (A => B) {
    private[this] def call[C](fn: A => C, a: A): TailRec[C] =
      fn match {
        case ref: CombineFunction1[A, _] =>
          for {
            lb <- tailcall(call(ref.left, a))
            rb <- tailcall(call(ref.right, a))
          } yield ref.semiB.combine(lb, rb)
        case _ => done(fn(a))
      }

    override def apply(a: A): B = call(this, a).result
  }
```